### PR TITLE
[LMA-2] Provide AlertManager-specific configuration base64 string

### PR DESCRIPTION
**Pull request by relaxdiego**
This change allows the charm user to provide AlertManager configuration
as base64-encoded strings. The rationale for this is so that systems
administrators who want to re-use an existing alertmanager.yml file can do so.
This also avoids the "abstraction over an abstraction" problem that would
be present if AlertManager-specific configuration was defined via the
charm's config.yml file. For example, AlertManager's inhibit_rules config
would have to be hand-massaged by the charm user to fit whatever is defined in
the charm's config.yaml file.

Additionl conversation is archived at: https://bit.ly/3fh0qNJ

### CODE REVIEW

**Comment by relaxdiego** [comment]
@exceptorr This is still draft and I have yet to update and add more unit tests. However, I wanted to get early feedback from you to know what you think of this approach.

**Comment by relaxdiego** [comment]
I think it makes sense to use `config.yaml` for configuring the charm but use `--resource` to pass the configuration for the app being managed by the charm.

**Comment by relaxdiego** [comment]
Another possible design choice is that the `alertmanager-config` and `alertmanager-secrets` resources are available for advanced charm users but for beginner or intermediate users, a limited set of configuration options is available via the charm's `config.yml`. Let's call these two scenarios **ADVANCED configuration** and **SIMPLIFIED configuration** respectively.

When using the **ADVANCED** configuration, the charm will exclusively use the files given via `alertmanager-config` and, when present, `alertmanager-secrets`.

When using the **SIMPLIFIED** configuration, the charm will first load a static alertmanager config file with sane defaults and then overwrite some of the appropriate settings with what is given by the charm user via the `--config` option.

**Comment by relaxdiego** [comment]
Example `juju deploy` command

```
juju deploy . \
  --resource alertmanager-config=$HOME/.alertmanager/config.yml \
  --resource alertmanager-secrets=$HOME/.alertmanager/secrets.yml \
  --resource alertmanager-image=prom/alertmanager:v0.20.0 \
  --config global_smtp_from='foo@bar.com'
```

`$HOME/.alertmanager/config.yml` and `$HOME/.alertmanager/secrets.yml` would be structured according to the official [AlerManager Configuration Schema](https://prometheus.io/docs/alerting/configuration/).

**Comment by relaxdiego** [comment]
Cross referencing [a Discourse thread that I started earlier](https://discourse.juju.is/t/how-to-reconcile-config-yamls-deliberate-restrictiveness-with-hierarchical-app-config/3184?u=mmaglana).

**Comment by exceptorr** [comment]
Why can't we simplify this and implement the Alertmanager configuration using direct YAML injection, rather than playing around with the resources?

The problem here is that you won't be able to supply a simple "just deploy it" bundle, you'll have to provide the resource files as well; while by giving an option to supply YAML manually you're giving some freedom to the charm operator AND allowing him to use basic IaC concepts by exporting his configs via `juju export-bundle`, for example.

**Comment by relaxdiego** [comment]
> Why can't we simplify this and implement the Alertmanager configuration using direct YAML injection, rather than playing around with the resources?

I’m interested. What would the CLI command look like in this case?

**Comment by exceptorr** [comment]
`juju config alertmanager config-file=$(base64 alertmanager-config.yaml)` for example (the same approach was taken for the ssl-* config options in the OpenStack charms)

**Comment by exceptorr** [comment]
Or, `include-base64:///home/ubuntu/alertmanager-config.yaml` in the bundle, if `juju deploy ...` is invoked.

**Comment by exceptorr** [comment]
Well, technically, we don't even have to use base64 - I don't know an exact reason why b64 encoding was chosen (probably, to avoid formatting and whitespace problems) - but the approach will remain the same anyway.

**Comment by relaxdiego** [comment]
@exceptorr: Alright, for my next pass I'm going to exclusively use the `--config` method you described even though [this documentation discourages it](https://discourse.juju.is/t/how-to-reconcile-config-yamls-deliberate-restrictiveness-with-hierarchical-app-config/3184) because your argument makes a lot of sense from a user experience perspective. 

I will also make use of [Juju 2.7's support for k8s secrets](https://discourse.juju.is/t/updated-podspec-yaml-new-features/2124) to turn the config file into a k8s secret before mounting onto the container.

File: src/adapters/framework.py [hunk 1]
```diff
@@ -58,6 +61,25 @@ def _fetch_image_meta(image_name, resources_repo):
...
+
+    if not path.exists():
+        msg = "Resource not found at {}".format(path)
+        logger.error(msg)
+        raise ResourceError(resource_name, msg)
+
+    logger.debug("Resource found at {}".format(path))
+    return path.read_text()
+
+
```
Comment (relaxdiego) [inline src/adapters/framework.py]: Dead code. Needs to be removed.

File: src/domain.py [hunk 2]
```diff
@@ -124,24 +145,63 @@ def to_json(self):
...
+
+        logger.debug("Loading secrets_yaml to dict")
+        secrets_dict = yaml.safe_load(secrets_yaml)
+
+        logger.debug("Updating AlertManager configuration with secrets")
+        alertmanager_config.update(secrets_dict)
+    else:
+        logger.info("alertmanager-secrets not provided. Ignoring")
+
+    # TODO: Remove this before merging as it exposes sensitive config info
```
Comment (exceptorr) [inline src/domain.py]: Remove this? :-)

**Review by exceptorr — CHANGES_REQUESTED** [review]
Looks good, one small comment and we're good to go!

File: src/domain.py (hunk 2 above)
Comment (relaxdiego) [reply src/domain.py]: LOL yes.

**Comment by relaxdiego** [comment]
@exceptorr being that you've given your conditional approval and that i've addressed it, I'm going to go ahead and merge this baby.

**Pull request merged**